crypto/fipsmodule/sha/sha3.c

```diff
@@ -377,3 +377,48 @@ int SHAKE_Squeeze(uint8_t *md, KECCAK1600_CTX *ctx, size_t len) {
   //FIPS_service_indicator_update_state();
   return 1;
 }
+
+int SHAKE128_Init_x4(KECCAK1600_CTX_x4 *ctx) {
+
+  int ok = (SHAKE_Init(&(*ctx)[0], SHAKE128_BLOCKSIZE) &&
+            SHAKE_Init(&(*ctx)[1], SHAKE128_BLOCKSIZE) &&
+            SHAKE_Init(&(*ctx)[2], SHAKE128_BLOCKSIZE) &&
+            SHAKE_Init(&(*ctx)[3], SHAKE128_BLOCKSIZE));
+
+  return ok;
+}
+
+int SHAKE128_Absorb_once_x4(KECCAK1600_CTX_x4 *ctx, const void *data0, const void *data1,
+                                  const void *data2, const void *data3, size_t len) {
+
+  int ok = (SHAKE_Absorb(&(*ctx)[0], data0, len) &&
+            SHAKE_Absorb(&(*ctx)[1], data1, len) &&
+            SHAKE_Absorb(&(*ctx)[2], data2, len) &&
+            SHAKE_Absorb(&(*ctx)[3], data3, len));
+
+  return ok;
+}
+
+int SHAKE128_Squeezeblocks_x4(uint8_t *md0, uint8_t *md1, uint8_t *md2, uint8_t *md3,
```

Comment:
Do we need the generality of an arbitrary `len`, or can we stick to a multiple of the block size for now? I seem to remember that ML-KEM only squeezes a block a time.

Reply:
If the generality of an arbitary length is needed, the function should not be called `Squeezeblocks`.

Reply:
SHAKE128_Squeezeblocks_x4 aims at defining the new API needed by mlkem-native. Thus, it is defined as mlkem-specific function (only allows squeezes of entire blocks), while _currently_ internally calling existing SHAKE sequential APIs (to split changes into incremental updates).
That is my understanding based on our previous discussions.
The same reasoning is applied to the SHAKE128_Absorb_once_x4 mlkem-specific API which inherits the generic functionality from the underlying existing SHAKE functions.

Reply:
My point is that the current implementation does not squeeze a number of blocks, but `len` is used as the number of bytes?

Reply:
I think you want something like (but 4-fold) https://github.com/aws/aws-lc/blob/09ad19eb3290157c666953a9d455efd0fed3fced/crypto/fipsmodule/ml_kem/fips202_glue.h#L36-L41

Reply:
Thank you! I just updated the code.